### PR TITLE
Optimize autocomplete. Fixes #589

### LINKED ISSUE #589: matcher.rs#search_frecent is very slow
Together with @csadilek we sat down to figure out what's causing https://github.com/mozilla-mobile/reference-browser/issues/395, and we narrowed it down to regressive cases of search_frecent.

Happy case is where we hit the requested limit of results (https://github.com/mozilla/application-services/issues/322) and stop asking matchers for results. We're seeing fast-enough performance here. On my test device (Nexus 6P) `search_frecent` call which is able to bail out early (because it accumulated enough results) is taking around 20-30ms.

In a regressive case when there are either no results, or relatively little (under the limit), `search_frecent` is taking around 10x the happy case. I've seen timings go into 300ms territory on Nexus 6P.

This slowness is then causing problems upstream, and we end up with a very sluggish interface when user is typing in a URL they've never visited before, for example. Reported delays are 1-2s between key strokes.

I'm timing call to `match_with_limit` specifically, not the individual matchers. Currently I don't know if it's a single matcher that's causing a slowdown, or if it's a death by a five papercuts.

We've also observed a cascading slow-down effect if the user is typing quickly, with the delay growing somewhere in the JNA machinery. Rust calls + JNA processing may take up to 700-1000ms in _very_ regressive cases. JNA processing goes back to normal if you slow down the rate of calls. Not yet sure what's causing this, but I think if we fix the underlying slowness it won't be a problem. Time it takes for the underlying `match_with_limit` to complete is mostly a constant and function of its input.

### PR DESCRIPTION
**Pull request by thomcc**
This makes it ~4x faster on my machine. About half of it comes from only running autocomplete queries a single time, and the other half from the missing index. The rest of the things do speed things up by about 100ms though (on my 160k row test case).

This manages to avoid [optimizing the hell out of AUTOCOMPLETE_MATCH](https://github.com/thomcc/application-services/commit/f36aeeab46d69f2ead48c75e1e6b56ce9bf93d89), since while that code does speed things up a bit, it's also a lot harder to maintain.

Important: it's still too slow to block the UI thread. @grigoryk / @csadilek really, you need to run this in the background asynchronously without blocking the user's typing on the response. (Even our autocomplete demo code manages to do this!)

That said, it is a lot better and closes the gap with desktop.

### CODE REVIEW

File: components/places/src/api/matcher.rs [hunk 1]
```diff
@@ -35,12 +35,7 @@ pub fn search_frecent(conn: &PlacesDb, params: SearchParams) -> Result<Vec<Searc
         &[
             // Try to match on the origin, or the full URL.
             &OriginOrUrl::new(&params.search_string, conn),
-            // After the first result, try the queries for adaptive matches and
-            // suggestions for bookmarked URLs.
-            &Adaptive::new(&params.search_string, conn),
-            &Suggestions::new(&params.search_string, conn),
-            // If we don't have enough results, query adaptive matches and
-            // suggestions again, matching anywhere instead of on boundaries.
+            // query adaptive matches and suggestions again, matching anywhere.
```
Comment (linacambridge) [inline components/places/src/api/matcher.rs]: Comment nit: IIUC, we don't query these "again" anymore; we just use `Anywhere`, and don't try matching on boundaries at all.

File: components/places/src/api/matcher.rs [hunk 2]
```diff
@@ -108,18 +104,12 @@ pub fn split_after_prefix(href: &str) -> (&str, &str) {
...
-        if c == '@' {
-            start = index + 1;
-        }
-    }
-    (&remainder[start..end], &remainder[end..])
+    let start = memchr::memchr(b'@', remainder.as_bytes())
+        .map(|i| i + 1)
+        .unwrap_or(0);
+    let remainder = &remainder[start..];
+    let end = memchr::memchr3(b'/', b'?', b'#', remainder.as_bytes()).unwrap_or(remainder.len());
```
Comment (linacambridge) [inline components/places/src/api/matcher.rs]: I was curious if there's a `memchr4`, `memchr5`, etc. 😄 BurntSushi/rust-memchr#11 was a fun read.

File: components/places/src/db/db.rs [hunk 3]
```diff
@@ -131,71 +127,63 @@ impl Deref for PlacesDb {
...
+        if raw == ValueRef::Null {
+            return Ok(None);
+        }
+        Ok(Some(raw.as_str().map_err(|e| {
+            Error::UserFunctionError(format!("Bad arg {} to '{}': {}", idx, fname, e).into())
+        })?))
+    }
+
+    // Note: The compiler can't meaningfully inline these, but if we don't put
+    // #[inline(never)] on them they get "inlined" into a temporary Box<FnMut>,
```
Comment (linacambridge) [inline components/places/src/db/db.rs]: Ewww... But thanks for adding the comment explaining it!

File: components/places/src/db/schema.rs [hunk 4]
```diff
@@ -178,7 +178,7 @@ const CREATE_TABLE_META_SQL: &str = "CREATE TABLE moz_meta (
 // See https://searchfox.org/mozilla-central/source/toolkit/components/places/nsPlacesIndexes.h
 const CREATE_IDX_MOZ_PLACES_URL_HASH: &str = "CREATE INDEX url_hashindex ON moz_places(url_hash)";
 
-// const CREATE_IDX_MOZ_PLACES_REVHOST: &str = "CREATE INDEX hostindex ON moz_places(rev_host)";
+const CREATE_IDX_MOZ_ORIGINS_REVHOST: &str = "CREATE INDEX hostindex ON moz_origins(rev_host)";
```
Comment (linacambridge) [inline components/places/src/db/schema.rs]: 👍

File: components/places/src/db/db.rs (hunk 3 above)
Comment (thomcc) [reply components/places/src/db/db.rs]: It's not super important tbh, but leaving them off saves us basically nothing, and I ended up adding them every time I went to profile this.

Comment (linacambridge) [reply components/places/src/db/db.rs]: Oh, the "ewww" was directed at the compiler, not your code! 😅

File: components/places/src/db/schema.rs [hunk 5]
```diff
@@ -245,9 +245,33 @@ pub fn init(db: &PlacesDb) -> Result<()> {
     Ok(())
 }
 
+/// Helper for upgrade. Intended use:
+///
+/// ```rust,ignore
+/// from = migration(db, from, 2, &[stuff, to, migrate, version2, to, version3])?;
```
Comment (linacambridge) [inline components/places/src/db/schema.rs]: This makes sense, but threw me off for a sec; I wondered why `migration` checked `cur_version == from` ("isn't 
the point of migration to migrate from an older version to a newer one"...but I misunderstood how it was meant to be used). What do you think of adding something like this?

```suggestion
/// let from = get_current_schema_version()?; // 2
/// from = migration(db, from, 2, &[stuff, to, migrate, version2, to, version3])?;
```

File: components/places/src/db/schema.rs [hunk 6]
```diff
@@ -245,9 +245,33 @@ pub fn init(db: &PlacesDb) -> Result<()> {
...
+/// from = migration(db, from, 3, &[stuff, to, migrate, version3, to, version4])?;
+/// from = migration(db, from, 4, &[stuff, to, migrate, version4, to, version5])?;
+/// ```
```
Comment (linacambridge) [inline components/places/src/db/schema.rs]: ...And here:

```suggestion
/// assert_eq!(get_current_schema_version(), 5);
/// ```
```

File: components/places/src/db/schema.rs [hunk 7]
```diff
@@ -245,9 +245,33 @@ pub fn init(db: &PlacesDb) -> Result<()> {
...
+///
+/// Etc.
+fn migration(db: &PlacesDb, cur_version: i64, from: i64, stmts: &[&str]) -> Result<i64> {
```
Comment (linacambridge) [inline components/places/src/db/schema.rs]: We might eventually want this to run more complicated migration functions, not just statements, but this looks great now. We can generalize it later if we need.

File: components/places/src/db/schema.rs [hunk 8]
```diff
@@ -245,9 +245,33 @@ pub fn init(db: &PlacesDb) -> Result<()> {
...
+        Ok(cur_version)
+    }
+}
+
 // https://github.com/mozilla-mobile/firefox-ios/blob/master/Storage/SQL/LoginsSchema.swift#L100
-fn upgrade(_db: &PlacesDb, from: i64) -> Result<()> {
+fn upgrade(db: &PlacesDb, mut from: i64) -> Result<()> {
     log::debug!("Upgrading schema from {} to {}", from, VERSION);
+
+    from = migration(db, from, 2, &[CREATE_IDX_MOZ_ORIGINS_REVHOST])?;
```
Comment (linacambridge) [inline components/places/src/db/schema.rs]: We should run this (all of schema initialization, really...maybe https://github.com/mozilla/application-services/blob/7981add8357c35aca6e21ac64503d97ec0b74c51/components/places/src/db/db.rs#L84 is a better place) in a transaction.

Comment (thomcc) [reply components/places/src/db/schema.rs]: I don't think the PRAGMA user_version updates are transacted on :(

**Pull request merged**